providers/common/kdfs/scrypt.c

```diff
@@ -13,268 +13,215 @@
+static void kdf_scrypt_init(KDF_SCRYPT *ctx);
 
-    while (*nptr) {
-        unsigned int digit;
-        uint64_t new_value;
-
-        if ((*nptr < '0') || (*nptr > '9')) {
-            return 0;
-        }
-        digit = (unsigned int)(*nptr - '0');
-        new_value = (value * 10) + digit;
-        if ((new_value < digit) || ((new_value - digit) / 10 != value)) {
-            /* Overflow */
-            return 0;
-        }
-        value = new_value;
-        nptr++;
-    }
-    *result = value;
-    return 1;
-}
-
-static EVP_KDF_IMPL *kdf_scrypt_new(void)
+static void *kdf_scrypt_new(void *provctx)
 {
-    EVP_KDF_IMPL *impl;
+    KDF_SCRYPT *ctx;
 
-    impl = OPENSSL_zalloc(sizeof(*impl));
-    if (impl == NULL) {
-        KDFerr(KDF_F_KDF_SCRYPT_NEW, ERR_R_MALLOC_FAILURE);
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
         return NULL;
     }
-    kdf_scrypt_init(impl);
-    return impl;
+    ctx->provctx = provctx;
+    kdf_scrypt_init(ctx);
+    return ctx;
 }
 
-static void kdf_scrypt_free(EVP_KDF_IMPL *impl)
+static void kdf_scrypt_free(void *vctx)
 {
-    kdf_scrypt_reset(impl);
-    OPENSSL_free(impl);
+    KDF_SCRYPT *ctx = (KDF_SCRYPT *)vctx;
+
+    kdf_scrypt_reset(ctx);
+    OPENSSL_free(ctx);
 }
 
-static void kdf_scrypt_reset(EVP_KDF_IMPL *impl)
+static void kdf_scrypt_reset(void *vctx)
 {
-    OPENSSL_free(impl->salt);
-    OPENSSL_clear_free(impl->pass, impl->pass_len);
-    memset(impl, 0, sizeof(*impl));
-    kdf_scrypt_init(impl);
+    KDF_SCRYPT *ctx = (KDF_SCRYPT *)vctx;
+
+    OPENSSL_free(ctx->salt);
+    OPENSSL_clear_free(ctx->pass, ctx->pass_len);
+    memset(ctx, 0, sizeof(*ctx));
+    kdf_scrypt_init(ctx);
 }
 
-static void kdf_scrypt_init(EVP_KDF_IMPL *impl)
+static void kdf_scrypt_init(KDF_SCRYPT *ctx)
 {
     /* Default values are the most conservative recommendation given in the
      * original paper of C. Percival. Derivation uses roughly 1 GiB of memory
      * for this parameter choice (approx. 128 * r * N * p bytes).
      */
-    impl->N = 1 << 20;
-    impl->r = 8;
-    impl->p = 1;
-    impl->maxmem_bytes = 1025 * 1024 * 1024;
+    ctx->N = 1 << 20;
+    ctx->r = 8;
+    ctx->p = 1;
+    ctx->maxmem_bytes = 1025 * 1024 * 1024;
 }
 
 static int scrypt_set_membuf(unsigned char **buffer, size_t *buflen,
-                             const unsigned char *new_buffer,
-                             size_t new_buflen)
+                             const OSSL_PARAM *p)
 {
-    if (new_buffer == NULL)
+    if (p->data == NULL)
         return 1;
+    if (p->data_size == 0) {
+        if ((*buffer = OPENSSL_malloc(1)) == NULL) {
+            ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+            return 0;
+        }
+    } else {
+        OPENSSL_clear_free(*buffer, *buflen);
+        *buffer = NULL;
+        if (!OSSL_PARAM_get_octet_string(p, (void **)buffer, 0, buflen))
+            /* TODO: raise an error here? */
```

Comment:
?

Reply:
This has the same answer because it is a copy/paste.

Reply:
Fixed.